MaterialSkin/Plugin.pm

```diff
@@ -2024,22 +2072,77 @@ sub _cliCommand {
             my $cnt = 0;
             foreach my $item ( @{ $req->getResult('playlists_loop') || [] } ) {
                 if ($cnt<NUM_HOME_ITEMS) {
-                    foreach my $key (keys(%{$item})) {
-                        my $val = $item->{$key};
-                        $request->addResultLoop("material_home_playlists_loop", $cnt, ${key}, ${val});
-                    }
-                    $request->addResultLoop("material_home_playlists_loop", $cnt, "ihe", 1);
+                    _addExtraHomeItem($request, "playlists", $item, $cnt);
                 }
                 $cnt+=1;
             }
             $request->addResult("material_home_playlists_loop_len", $cnt);
         }
+
+        my $others = [ grep { $_ } map { getHomeExtra($_) } keys %{$request->getParamsCopy()} ];
+        if (scalar @$others) {
+            $request->setStatusProcessing();
+
+            # process other home items asynchronously and in parallel - they might be doing online lookups
+            Async::Util::amap(
+                inputs => $others,
+                action => sub {
+                    my ($extra, $acb) = @_;
+                    my $id = $extra->{id};
+
+                    $extra->{handler}->($request->client, sub {
+                        $acb->({ $id => (shift || []) });
+                    });
+                },
+                at_a_time => 4,
+                cb => sub {
+                    my $results = {};
+
+                    foreach my $result (@{$_[0]}) {
+                        my ($id, $res) = each %{$result};
+                        $results->{$id} = $res;
+                    }
+
+                    foreach my $id (map { $_->{id} } sort { $a->{weight} <=> $b->{weight} } @$others) {
+                        my $cnt = 0;
+
+                        next unless $results->{$id};
+
+                        foreach my $item ( @{$results->{$id}} ) {
+                            last if $cnt >= NUM_HOME_ITEMS;
+
+                            _addExtraHomeItem($request, $id, $item, $cnt);
+                            $cnt++;
+                        }
+
+                        $request->addResult("material_home_${id}_loop_len", scalar @{$results->{$id}});
+                    }
+
+                    $request->setStatusDone();
+                }
+            );
+
+            # we have to wait for the async processing to finish
+            return;
+        }
+
         $request->setStatusDone();
         return;
     }
     $request->setStatusBadParams();
 }
 
+sub _addExtraHomeItem {
+    my ($request, $id, $item, $cnt) = @_;
+
+    foreach my $key (keys(%{$item})) {
+        my $val = $item->{$key};
+        $request->addResultLoop("material_home_${id}_loop", $cnt, ${key}, ${val});
+    }
+
+    $request->addResultLoop("material_home_${id}_loop", $cnt, "ihe", 1);
```

Comment:
BTW: I have no clue what the `ihe` is for here. I stole this from your code...

Reply:
Means it's ok to leave it as it is?

Reply:
Yeah, just leave.

One further question. There will need to be some property that holds the command to fetch the complete list (limited to new music length) when "More" is clicked.